[ALGO]Bron-Kerbosch with pivoting

Hromz opened the pull request:
Added algorithm, test and documentation.
In the end I got two implementations of the algorithm, but I decided to stick with the unordered_set implementation. Second one was bitwise approach with ULL implementation, unfortunately it didn't pass the runtime test (I guess I missed something in the implementation). In the future I would like to add bitwise approach and for sparse graph with vertex ordering appraoch.

Looking forward for your comments and suggestions!

closes #95

bobluppes commented on include/graaflib/algorithm/clique/bron_kerbosch.tpp:
```diff
@@ -0,0 +1,97 @@
+  auto union_set = vertices;
+  for (const auto& vertex : excluded_vertices) union_set.insert(vertex);
+
+  // Get pivot vertex with most amount of neighbors
+  // to reduce recursion calls
+  auto pivot_vertex = *union_set.begin();
+  for (const auto& vertex : union_set) {
+    if (graph.get_neighbors(pivot_vertex).size() <
+        graph.get_neighbors(vertex).size()) {
+      pivot_vertex = vertex;
+    }
+  }
```
This works and should be fine to leave as-is. However, if you are up for it this sounds like a great use-case to use C++ 20's [std::ranges::max_element](https://en.cppreference.com/w/cpp/algorithm/ranges/max_element) in combination with a projection.

```c++
const auto pivot_vertex = *std::ranges::max_element(union_set, {}, [&graph](vertex_id_t id) {
  return graph.get_neighbors(id).size();
};
```

bobluppes commented on include/graaflib/algorithm/clique/bron_kerbosch.tpp:
```diff
@@ -0,0 +1,97 @@
+    }
+  }
+
+  // vertices_to_process = candidate_vertices \ N(pivot_vertex)
+  auto vertices_to_process = vertices;
+  for (const auto& vertex : graph.get_neighbors(pivot_vertex)) {
+    vertices_to_process.erase(vertex);
+  }
```
Same here, might be a nice idea to extract this to a set difference function. i.e.

```c++
std::unordered_set<vertex_id_t> set_difference(const std::unordered_set<vertex_id_t>& lhs, const std::unordered_set<vertex_id_t> rhs) {...}
```

bobluppes commented on test/graaflib/algorithm/clique/bron_kerbosch_test.cpp:
```diff
@@ -0,0 +1,299 @@
+  using graph_t = T;
+};
+
+using graph_types = testing::Types<undirected_graph<int, int>>;
+TYPED_TEST_SUITE(BronKerbosch, graph_types);
```
[minor]: If we only have one graph type we are testing, there is no real need for a `TYPED_TEST_SUITE`, instead we could just write the type directly in the tests. However, if you do want to test this for different graph types, we now have predefined type collections you can use here in `<utils/fixtures/fixtures.h>`.

bobluppes commented on test/graaflib/algorithm/clique/bron_kerbosch_test.cpp:
```diff
@@ -0,0 +1,299 @@
+  ASSERT_EQ(cliques, expected_cliques);
+}
+
+TYPED_TEST(BronKerbosch, RandomGraphCheckRunTime) {
+  using graph_t = typename TestFixture::graph_t;
+  graph_t graph{};
+
+  std::vector<vertex_id_t> vertices;
+
+  for (int i = 1; i <= 70; ++i) {
+    vertices.push_back(graph.add_vertex(i));
+  }
+
+  std::random_device dev;
+  std::mt19937 rng(dev());
+  std::uniform_int_distribution<std::mt19937::result_type> rand_vertex(0, 69);
+
+  for (int i = 0; i < 1360; ++i) {
+    graph.add_edge(vertices[rand_vertex(rng)], vertices[rand_vertex(rng)], 1);
+  }
+
+  auto start = std::chrono::high_resolution_clock::now();
+  auto cliques = bron_kerbosch(graph);
+  auto stop = std::chrono::high_resolution_clock::now();
+
+  int expected_time = 900;
+
+  ASSERT_TRUE(expected_time >=
+              duration_cast<std::chrono::milliseconds>(stop - start).count());
+}
```
Great idea to measure the runtime performance 👍🏻 

Not sure if you saw, but we are also using the google benchmark framework in `/perf`, where we get nice functionalities such as repeated executions of the algorithm after which the average runtime (and memory usage) is reported. There we can also nicely parametrize the benchmark to test for increasingly large graphs.

Such a benchmark might be a better suited place to measure runtime performance of our algorithms than in unit tests. As such I would propose to remove this unit test and to create a follow-up issue in which we can add the benchmark.

bobluppes reviewed (commented):
Hi @Hromz, as always great work 🚀 

Hereby some feedback, please let me know what you think.

I really like how you extracted some of the set operations (intersection) into a function. It is super unfortunate that C++ or the stl does not give this for free. In a follow-up, it would be cool if we can extract such set operations to free functions and put them in a central place in the library which we can use as an extension to the stl. Out of scope for this PR but just some food for thought :)

Hromz commented:
Hey @bobluppes! Thanks for your comments, with each PR I'm discovering something new. :)  I corrected the code according to your comments. Renamed folder from "Clique" to "Clique-detection". Deleted the runtime test and added two more tests.

Hromz commented on README.md:
```diff
@@ -92,24 +92,27 @@ Algorithms implemented in the Graaf library include the following. For more info
+    - Dijkstra
+    - Bellman-Ford
+5. [**Strongly Connected Components Algorithms
+   **](https://bobluppes.github.io/graaf/docs/category/strongly-connected-components):
```
Hmm, think i pressed auto alignment by mistake. Corrected it.

bobluppes reviewed (approved):
Thanks for addressing, looks excellent 🎉
